src/Features/Core/Portable/MakeFieldReadonly/MakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,254 @@
+        {
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                // State map for fields:
+                //  'isCandidate' : Indicates whether the field is a candidate to be made readonly based on it's declaration and options.
+                //  'written'     : Indicates if there are any writes to the field outside the constructor and field initializer.
+                var fieldStateMap = new ConcurrentDictionary<IFieldSymbol, (bool isCandidate, bool written)>();
+
+                // We register following actions in the compilation:
+                // 1. A symbol action for field symbols to ensure the field state is initialized for every field in the compilation.
+                // 2. An operation action for field references to detect if a candidate field is written outside constructor and field initializer, and update field state accordingly.
+                // 3. A symbol start/end action for named types to report diagnostics for candidate fields that were not written outside constructor and field initializer.
+
+                compilationStartContext.RegisterSymbolAction(symbolContext =>
+                {
+                    _ = TryGetOrInitializeFieldState((IFieldSymbol)symbolContext.Symbol, symbolContext.Options, symbolContext.CancellationToken);
+                }, SymbolKind.Field);
+
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    symbolStartContext.RegisterOperationAction(operationContext =>
+                    {
+                        var fieldReference = (IFieldReferenceOperation)operationContext.Operation;
+                        (bool isCandidate, bool written) = TryGetOrInitializeFieldState(fieldReference.Field, operationContext.Options, operationContext.CancellationToken);
+
+                        // Ignore fields that are not candidates or have already been written outside the constructor/field initializer.
+                        if (!isCandidate || written)
+                        {
+                            return;
+                        }
+
+                        // Check if this is a field write outside constructor and field initializer, and update field state accordingly.
+                        if (IsFieldWrite(fieldReference, operationContext.ContainingSymbol))
+                        {
+                            UpdateFieldStateOnWrite(fieldReference.Field);
+                        }
+                    }, OperationKind.FieldReference);
+
+                    symbolStartContext.RegisterSymbolEndAction(symbolEndContext =>
```

Comment:
:memo: One of my primary objectives in the API review will be ensuring that these callbacks can be registered with incurring allocations for the callback delegates. The existing APIs do not support this in all cases, but we've observed it as a problem for a while and it will be helpful to not be limited in the same manner for new APIs.

Reply:
i'm curious in what cases these callbacks are a problem.  I would expect, normally, that any registration would just be a single allocatoin for the entire run of the analyzer.